### DISCUSSION TITLE
Windows configure script eats path separator?

### DISCUSSION THREAD

**Discussion by mchamberland** [nrc-cnrc/EGSnrc#1064] [Questions]
I'm trying to configure EGSnrc on a Windows 10 workstation. I'm using MinGW 7.5 from [here](https://winlibs.com/). 

During compilation of the egs++ library, certain paths are missing their separator so certain components are not properly built. Not everything, just certain ones, weirdly.

See part of the log:

```
cd geometry\egs_planes && mingw32-make
cd geometry\egs_cd_geometry && mingw32-make
cd geometry\egs_gtransformed && mingw32-make
cd geometry\egs_nd_geometry && mingw32-make
cd geometry\egs_box && mingw32-make
cd geometry\egs_genvelope && mingw32-make
cd geometry\egs_spheres && mingw32-make
cd geometry\egs_cylinders && mingw32-make
cd geometry\egs_iplanes && mingw32-make
cd geometry\egs_cones && mingw32-make
cd geometry\egs_gstack && mingw32-make
cd geometry\egs_prism && mingw32-make
cd geometry\egs_union && mingw32-make
cd geometry\egs_pyramid && mingw32-make
cd geometry\egs_conez && mingw32-make
cd geometry\egs_space && mingw32-make

/usr/bin/sh: line 0: cd: geometryegs_planes: No such file or directory
cd geometry\egs_elliptic_cylinders && mingw32-make

mingw32-make: *** [makefile:187: egs_planes] Error 1

mingw32-make: *** Waiting for unfinished jobs....

/usr/bin/sh: line 0: cd: geometryegs_nd_geometry: No such file or directory
/usr/bin/sh: line 0: cd: geometryegs_cd_geometry: No such file or directory

mingw32-make: *** [makefile:187: egs_nd_geometry] Error 1

mingw32-make: *** [makefile:187: egs_cd_geometry] Error 1

/usr/bin/sh: line 0: cd: geometryegs_spheres: No such file or directory
/usr/bin/sh: line 0: cd: geometryegs_box: No such file or directory

mingw32-make: *** [makefile:187: egs_box] Error 1

mingw32-make: *** [makefile:187: egs_spheres] Error 1

/usr/bin/sh: line 0: cd: geometryegs_genvelope: No such file or directory
/usr/bin/sh: line 0: cd: geometryegs_gtransformed: No such file or directory

/usr/bin/sh: line 0: cd: geometryegs_iplanes: No such file or directory
mingw32-make: *** [makefile:187: egs_genvelope] Error 1

mingw32-make: *** [makefile:187: egs_gtransformed] Error 1

/usr/bin/sh: line 0: cd: geometryegs_cones: No such file or directory
mingw32-make: *** [makefile:187: egs_iplanes] Error 1

/usr/bin/sh: line 0: cd: geometryegs_gstack: No such file or directory

mingw32-make: *** [makefile:187: egs_cones] Error 1

mingw32-make: *** [makefile:187: egs_gstack] Error 1

/usr/bin/sh: line 0: cd: geometryegs_union: No such file or directory

/usr/bin/sh: line 0: cd: geometryegs_prism: No such file or directory
mingw32-make: *** [makefile:187: egs_union] Error 1

/usr/bin/sh: line 0: cd: geometryegs_cylinders: No such file or directory
/usr/bin/sh: line 0: cd: geometryegs_space: No such file or directory
/usr/bin/sh: line 0: cd: geometryegs_elliptic_cylinders: No such file or directory
mingw32-make: *** [makefile:187: egs_cylinders] Error 1

/usr/bin/sh: line 0: cd: geometryegs_conez: No such file or directory
mingw32-make: *** [makefile:187: egs_space] Error 1

mingw32-make: *** [makefile:187: egs_conez] Error 1

/usr/bin/sh: line 0: cd: geometryegs_pyramid: No such file or directory
mingw32-make: *** [makefile:187: egs_prism] Error 1

mingw32-make: *** [makefile:187: egs_elliptic_cylinders] Error 1

mingw32-make: *** [makefile:187: egs_pyramid] Error 1
```

Any ideas what causes this? I'm a little out of my element on Windows...

**Comment by ftessier** [comment]
Weird, how can it be missing only in _some_ places? Is this through the GUI or command line?

**Reply by mchamberland** [reply]
> Weird, how can it be missing only in _some_ places? Is this through the GUI or command line?

Through the GUI. Not sure how to run the configure script on Windows from CMD.exe.

**Reply by mchamberland** [reply]
Also, Windows Subsystem for Linux is not an option, sadly.

**Comment by rtownson** [comment]
I'm not sure if compiler version is the issue, but the first thing I would try is switching to mingw 6.4.0 or 7.3.0 from here: https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/

Make sure to use a fully fresh install of EGSnrc if you switch compilers.

**Reply by mchamberland** [reply]
Same issue.

Interestingly, when I compile applications, I get something about unexpected end of line, but I can't figure out where this is coming from:

C:\data\egsnrc_clrp\egs_home\egs_app>mingw32-make
/usr/bin/sh: -c: line 1: syntax error: unexpected end of file
g++ -I. -IC:\data\egsnrc_clrp\HEN_HOUSE\lib\uvmmc -IC:\data\egsnrc_clrp\HEN_HOUSE\egs++ -IC:\data\egsnrc_clrp\HEN_HOUSE\interface -DWIN32 -fPIC  -O2 -mtune=native -c -o egs_app_uvmmc.o egs_app.cpp
g++ -I. -IC:\data\egsnrc_clrp\HEN_HOUSE\lib\uvmmc -IC:\data\egsnrc_clrp\HEN_HOUSE\egs++ -IC:\data\egsnrc_clrp\HEN_HOUSE\interface -DWIN32 -fPIC  -O2 -mtune=native -c -o egs_advanced_application_uvmmc.o C:\data\egsnrc_clrp\HEN_HOUSE\egs++\egs_advanced_application.cpp
Mortran compiling EGSnrc sources ...
gfortran -cpp -O2 -mtune=native -fPIC    -DCOMPILE_TIME="\"2023-09-28 02:26 PM Eastern Standard Time\"" -DGIT_HASH="\"c4beb16\"" -c -o egsnrc_uvmmc.o egsnrc_uvmmc.F
g++ -I. -IC:\data\egsnrc_clrp\HEN_HOUSE\lib\uvmmc -IC:\data\egsnrc_clrp\HEN_HOUSE\egs++ -IC:\data\egsnrc_clrp\HEN_HOUSE\interface -DWIN32 -fPIC  -O2 -mtune=native -c -o egs_interface2_uvmmc.o C:\data\egsnrc_clrp\HEN_HOUSE\interface\egs_interface2.c
g++ -O2 -mtune=native -o C:\data\egsnrc_clrp\egs_home\bin\uvmmc\egs_app.exe egs_app_uvmmc.o egs_advanced_application_uvmmc.o egsnrc_uvmmc.o egs_interface2_uvmmc.o -LC:\data\egsnrc_clrp\HEN_HOUSE\egs++\dso\uvmmc -legspp -lgfortran
g++ -I. -IC:\data\egsnrc_clrp\HEN_HOUSE\lib\uvmmc -IC:\data\egsnrc_clrp\HEN_HOUSE\egs++ -IC:\data\egsnrc_clrp\HEN_HOUSE\interface -DWIN32 -fPIC  -DBUILD_APP_LIB -O2 -mtune=native -c -o egs_app_lib_uvmmc.o egs_app.cpp
g++ -I. -IC:\data\egsnrc_clrp\HEN_HOUSE\lib\uvmmc -IC:\data\egsnrc_clrp\HEN_HOUSE\egs++ -IC:\data\egsnrc_clrp\HEN_HOUSE\interface -DWIN32 -fPIC  -DBUILD_APP_LIB -O2 -mtune=native -c -o egs_advanced_application_lib_uvmmc.o C:\data\egsnrc_clrp\HEN_HOUSE\egs++\egs_advanced_application.cpp
g++ -O2 -mtune=native -shared -LC:\data\egsnrc_clrp\HEN_HOUSE\egs++\dso\uvmmc egs_app_lib_uvmmc.o egs_advanced_application_lib_uvmmc.o egsnrc_uvmmc.o egs_interface2_uvmmc.o -o C:\data\egsnrc_clrp\egs_home\bin\uvmmc\egs_app.dll -Wl,--out-implib,C:\data\egsnrc_clrp\egs_home\bin\uvmmc\egs_app.lib  -legspp -lgfortran


Similarly in the egs++ folder:
C:\data\egsnrc_clrp\HEN_HOUSE\egs++>mingw32-make
/usr/bin/sh: -c: line 1: syntax error: unexpected end of file
mingw32-make: *** [Makefile:166: dso\uvmmc] Error 1


This is with MinGW 7.3.0.

Not sure if this is any helpful, but you can see in the egs++/dso folder that it's missing anything that is one folder or more folders deeper than the base egs++ folder (i.e., the geometry, shapes, and sources).
![image](https://github.com/nrc-cnrc/EGSnrc/assets/1675666/a7273a51-c6e1-4f01-a549-5e35939772f2)

But there is also a dso/win-static folder created with some of those missing dll.
![image](https://github.com/nrc-cnrc/EGSnrc/assets/1675666/ce2f94d0-5697-4ea7-af5c-f21563f9fda4)


Any ideas?

**Comment by mchamberland** [comment]
I was able to get the configure script running and egs_c_utils fails to compile. I'll look into that a bit more closely later. But in the mean time, I appreciate any suggestions!

**Reply by rtownson** [reply]
Is this on Windows 11? We don't have a Win11 system to test on as far as I know, so it might not work out of the box.

**Reply by mchamberland** [reply]
@rtownson no, this is Windows 10 Enterprise. It's a work computer, so severely locked down. Not the end of the world if I can't get it working. I will try on a workstation where I have more permissions in the meantime and see if I get any further.

**Reply by mainegra** [reply]
How did you run the script? WSL?

**Comment by mainegra** [comment]
Hi Marc,

Last time I did a Windows 10 install, in January 2023, I used version 8.1 and it worked just fine!

GNU Fortran (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0

It is also strange the folder separator is missing only in some cases as @ftessier pointed out.

**Comment by mainegra** [comment]
Interestingly, I just tried with GNU Fortran 8.3 (x86_64-posix-seh, Built by strawberryperl.com project) and it worked, except egs_phsp_scoring which fails with error:
```
egs_phsp_scoring.cpp: In constructor 'EGS_PhspScoring::EGS_PhspScoring(const string&, EGS_ObjectFactory*)':
egs_phsp_scoring.cpp:50:88: error: call of overloaded 'basic_fstream(int)' is ambiguous
     count(0), countg(0), emin(1.e30), emax(-1.e30), first_flush(true), is_restart(false) {


In file included from egs_phsp_scoring.cpp:38:
C:/Strawberry/c/lib/gcc/x86_64-w64-mingw32/8.3.0/include/c++/fstream:1101:7: note: candidate: 'std::basic_fstream<_CharT, _Traits>::basic_fstream(std::basic_fstream<_CharT, _Traits>&&) [with _CharT = char; _Traits = std::char_traits<char>]'
       basic_fstream(basic_fstream&& __rhs)
       ^~~~~~~~~~~~~
C:/Strawberry/c/lib/gcc/x86_64-w64-mingw32/8.3.0/include/c++/fstream:1099:7: note: candidate: 'std::basic_fstream<_CharT, _Traits>::basic_fstream(const std::basic_fstream<_CharT, _Traits>&) [with _CharT = char; _Traits = std::char_traits<char>]' <deleted>
       basic_fstream(const basic_fstream&) = delete;
       ^~~~~~~~~~~~~
C:/Strawberry/c/lib/gcc/x86_64-w64-mingw32/8.3.0/include/c++/fstream:1078:7: note: candidate: 'std::basic_fstream<_CharT, _Traits>::basic_fstream(const string&, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::__cxx11::string = std::__cxx11::basic_string<char>; std::ios_base::openmode = std::_Ios_Openmode]'
       basic_fstream(const std::string& __s,
       ^~~~~~~~~~~~~
C:/Strawberry/c/lib/gcc/x86_64-w64-mingw32/8.3.0/include/c++/fstream:1062:7: note: candidate: 'std::basic_fstream<_CharT, _Traits>::basic_fstream(const wchar_t*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::_Ios_Openmode]'
       basic_fstream(const wchar_t* __s,
       ^~~~~~~~~~~~~
C:/Strawberry/c/lib/gcc/x86_64-w64-mingw32/8.3.0/include/c++/fstream:1048:7: note: candidate: 'std::basic_fstream<_CharT, _Traits>::basic_fstream(const char*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::_Ios_Openmode]'
       basic_fstream(const char* __s,
       ^~~~~~~~~~~~~

```
